toolchains/idf/components/toit/CMakeLists.txt

```diff
@@ -0,0 +1,174 @@
+            "--envelope=firmware.envelope"
+            "create"
+            "--bootloader.bin=bootloader/bootloader.bin"
+            "--firmware.bin=${EXECUTABLE_NAME}.bin"
+            "--firmware.elf=${EXECUTABLE_NAME}.elf"
+            "--partitions.bin=partition_table/partition-table.bin"
+            "--partitions.csv=${CMAKE_HOME_DIRECTORY}/partitions.csv"
+            "--system.snapshot=${SYSTEM_SNAPSHOT}"
+            COMMAND echo envelope: created
+    )
+
+    if (CONFIG_TOIT_WIFI_SSID AND CONFIG_TOIT_WIFI_PASSWORD)
+        add_custom_command(
+                TARGET firmware
+                COMMAND "${FIRMWARE_BIN}"
+                "--envelope=firmware.envelope"
+                "property" "set" "wifi" "'{\"wifi.ssid\":\"${CONFIG_TOIT_WIFI_SSID}\",\"wifi.password\":\"${CONFIG_TOIT_WIFI_PASSWORD}\"}'"
+                COMMAND echo envelope: set wifi
+        )
+    endif()
+
+    if (CONFIG_TOIT_ENTRY_POINT)
+        set(RUN_PROGRAM true)
+        if (NOT EXISTS "${CONFIG_TOIT_ENTRY_POINT}")
+            message(WARNING "Could not find entry point: ${CONFIG_TOIT_ENTRY_POINT}")
+            set(RUN_PROGRAM false)
+        endif()
+
+        if (CONFIG_TOIT_ENTRY_POINT_PROJECT_ROOT)
+            if (NOT EXISTS "${CONFIG_TOIT_ENTRY_POINT_PROJECT_ROOT}")
+                message(WARNING "Could not find project root of entry point: ${CONFIG_TOIT_ENTRY_POINT_PROJECT_ROOT}")
+                set(RUN_PROGRAM false)
+            endif()
+            set(ENTRY_PROJECT_ROOT_ARGS "--project-root" "${CONFIG_TOIT_ENTRY_POINT_PROJECT_ROOT}")
+        endif()
+
+        if (RUN_PROGRAM)
+            add_custom_command(OUTPUT ${CMAKE_BINARY_DIR}/program.snapshot
+                    COMMAND "${TOITC}" "-w" "${CMAKE_BINARY_DIR}/program.snapshot" ${ENTRY_PROJECT_ROOT_ARGS} ${CONFIG_TOIT_ENTRY_POINT}
+                    VERBATIM)
+
+            add_custom_target(program_snapshot DEPENDS ${CMAKE_BINARY_DIR}/program.snapshot)
+            add_dependencies(firmware program_snapshot)
+
+            add_custom_command(
+                    TARGET firmware
+                    COMMAND "${FIRMWARE_BIN}"
+                    "--envelope=firmware.envelope"
+                    "container" "install" "program" ${CMAKE_BINARY_DIR}/program.snapshot
+                    COMMAND echo envelope: program added
+            )
+        endif()
+    endif()
+
+    add_custom_command(
+            TARGET firmware
+            COMMAND "${FIRMWARE_BIN}"
+            "--envelope=firmware.envelope"
+            "extract" "--firmware.bin" "-o" "${EXECUTABLE_NAME}-firmware.bin"
+            COMMAND echo envelope: firmware extracted
+    )
+
+    add_custom_target(mod_flash_args DEPENDS firmware)
+    add_dependencies(flash mod_flash_args)
+    add_custom_command(
+            TARGET mod_flash_args
+            COMMAND sed -i s/${EXECUTABLE_NAME}\\.bin/${EXECUTABLE_NAME}-firmware\\.bin/g ${CMAKE_BINARY_DIR}/flash_args
+            COMMAND echo MID: MOD FLASH ARGS
```

Comment:
What does this do?

Reply:
It convinces idf.py flash to use the correct updated binary file from the firmware tool. The echo is removed. A comment is added.